File: Week-1-C/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card = get_long("Number: ");
    int sum = 0, count = 0;
    long temp = card;
    long first_two = 0;

    while (temp > 0)
    {
        int digit = temp % 10;

        if (count % 2 == 0) // Every other starting from last
        {
            sum += digit;
        }
        else // Every other starting from second-to-last
        {
            int product = digit * 2;
            sum += (product / 10) + (product % 10);
        }

        if (temp >= 10 && temp < 100) first_two = temp; // Capture starting digits

        temp /= 10;
        count++;
    }

    // Checksum + Length + Starting Digits
    if (sum % 10 == 0)
    {
        if (count == 15 && (first_two == 34 || first_two == 37)) printf("AMEX\n");
        else if (count == 16 && (first_two >= 51 && first_two <= 55)) printf("MASTERCARD\n");
        else if ((count == 13 || count == 16) && (first_two / 10 == 4)) printf("VISA\n");
        else printf("INVALID\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week-6-Python/credit.py
from cs50 import get_int


def main():
    number = get_number()
    print(get_card_type(number))


def get_number():
    while True:
        n = get_int("Number: ")
        if n > 0:
            return n


def luhn_check(number):
    digits = [int(d) for d in str(number)]
    total = 0
    # Every second digit from the right, starting with the second-to-last
    for i, digit in enumerate(reversed(digits)):
        if i % 2 == 1:
            doubled = digit * 2
            total += doubled // 10 + doubled % 10
        else:
            total += digit
    return total % 10 == 0


def get_card_type(number):
    s = str(number)
    length = len(s)
    first_two = int(s[:2])
    first_one = int(s[0])

    if not luhn_check(number):
        return "INVALID"

    # AMEX: 15 digits, starts with 34 or 37
    if length == 15 and first_two in [34, 37]:
        return "AMEX"

    # MASTERCARD: 16 digits, starts with 51–55
    if length == 16 and 51 <= first_two <= 55:
        return "MASTERCARD"

    # VISA: 13 or 16 digits, starts with 4
    if length in [13, 16] and first_one == 4:
        return "VISA"

    return "INVALID"


main()
